wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -95,34 +163,191 @@ applyBlock pw@PassiveWallet{..} b = do
+                      Just rb -> Right rb
+
+      -- Compute the list of blocks that must be applied, along with the accounts that
+      -- should be updated for each block.
+      -- PRECONDITION: Each list of blocks should eminate from the same "newest" block.
+      gatherAcctsPerBlock :: NonEmpty (HdAccountId, OldestFirst [] ResolvedBlock)
+                          -> OldestFirst [] (ResolvedBlock, Set HdAccountId)
+      gatherAcctsPerBlock a2bs =
+        let firstAppearedIn :: Map Int (Set HdAccountId)
+            firstAppearedIn = Map.fromListWith Set.union
+                              $ map (\(a, bs) -> (longestLength - length bs, Set.singleton a))
+                              $ NE.toList a2bs
+
+            -- The longest sequence of blocks; due to the precondition on gatherAcctsPerBlock,
+            -- every sequence of blocks appearing in a2bs is a suffix of this sequence.
+            longest :: OldestFirst [] ResolvedBlock
+            longest = neMaximumBy (comparing length) $ map snd a2bs
+
+            longestLength :: Int
+            longestLength = length longest
+
+            neMaximumBy :: (a -> a -> Ordering) -> NonEmpty a -> a
+            neMaximumBy cmp = maximumBy cmp . NE.toList
+
+            -- The accounts to update for each block in 'longest'.
+            updateSets :: [Set HdAccountId]
+            updateSets = scanl' Set.union Set.empty
+                         $ map (\n -> Map.findWithDefault Set.empty n firstAppearedIn) [0..]
+        in OldestFirst $ zip (getOldestFirst longest) updateSets
+
 -- | Switch to a new fork
 --
 -- NOTE: The Ouroboros protocol says that this is only valid if the number of
 -- resolved blocks exceeds the length of blocks to roll back.
 switchToFork :: PassiveWallet
-             -> Int             -- ^ Number of blocks to roll back
+             -> Maybe HeaderHash -- ^ Roll back until we meet this hash.
              -> [ResolvedBlock] -- ^ Blocks in the new fork
-             -> IO (Either SwitchToForkError ())
-switchToFork pw@PassiveWallet{..} n bs = do
+             -> IO ()
+switchToFork pw@PassiveWallet{..} oldest bs = do
     k <- Node.getSecurityParameter _walletNode
     blocksAndMeta <- mapM (prefilterBlock' pw) bs
     let (blockssByAccount, metas) = unzip blocksAndMeta
-    res <- update' _wallets $ SwitchToFork k n blockssByAccount
-    case res of
-      Left  err     -> return $ Left err
-      Right changes -> do mapM_ (putTxMeta _walletMeta) $ concat metas
-                          modifyMVar_ _walletSubmission $
-                            return . Submission.addPendings (fst <$> changes)
-                          modifyMVar_ _walletSubmission $
-                            return . Submission.remPending (snd <$> changes)
-                          return $ Right ()
+
+    changes <- trySwitchingToFork k blockssByAccount
+
+    -- Update the metadata
+    mapM_ (putTxMeta _walletMeta) $ concat metas
+    modifyMVar_ _walletSubmission $
```

Comment:
@matt-noonan Unrelated (and minor): is it worthwhile to mush this into a single atomic update rather than having two different `modifyMVar_`?